universal-application-tool-0.0.1/app/services/program/BlockDefinition.java

```diff
@@ -111,13 +112,19 @@ public boolean isFileUpload() {
         .anyMatch(questionType -> questionType.equals(QuestionType.FILEUPLOAD));
   }
 
-  /** A {@link Predicate} that determines whether this is hidden or shown. */
+  /** A {@link PredicateDefinition} that determines whether this is hidden or shown. */
   @JsonProperty("hidePredicate")
-  public abstract Optional<Predicate> hidePredicate();
+  public abstract Optional<PredicateDefinition> predicate();
```

Comment:
question: Should the @JsonProperty name match this method name?

Related suggestion/question: Maybe we should keep the `hide` prefix (or change it to `show`? depending how we are doing it. I guess we haven't thought through the details of how we'll offer both "show if" and "hide if" to the admin user...). Removing the prefix allows us some flexibility (i.e., if `PredicateDefinition` has some bool or enum like `HIDE_IF`/`SHOW_IF`, then maybe this predicate *isn't* one or the other, but could represent either one). This is not a well-formed suggestion... Just bringing it up as a discussion/thought.

Reply:
Thanks Natalie, really good questions/comments! First, as far as I know we have to keep the "hidePredicate" json property because existing questions in the database use this field name - for example, this is a serialized block I pulled from staging's `/dev/seed`:

```json
"blockDefinitions" : [ {
    "id" : 1,
    "name" : "Block 1",
    "description" : "not translated",
    "repeaterId" : null,
    "hidePredicate" : null,
    "optionalPredicate" : null,
    "questionDefinitions" : [ {
      "id" : 6
    } ]
  } ],
```

For the method name, what would you think of something like `visibilityPredicate`? We currently have a plan to offer both `SHOW_BLOCK` and `HIDE_BLOCK` actions in `PredicateDefinition`, so this could capture both actions.